### PR TITLE
Warn on case mismatch while opening a file on macOS

### LINKED ISSUE #23441: macOS filesystems can be case insensitive, editor should also warn like Windows on case mismatch
I found an irritating bug that took a long time to track down. I was able to make a minimal reproducible project. The project would work when launching in Godot but crash immediately when launching on the exported Mac app.

Suppose we have a script like this:

```gdscript
# res://base/Bar.gd

extends Node
```

and another file

```gdscript
# res://Foo.gd

extends 'res://Base/Bar.gd'
```

Notice the capitalization mistake? Neither did I. Godot, I guess, will ignore capitalization in directories. The project runs just fine. It throws no errors. But when exporting to Mac, it will not launch. Caused me a lot of frustration.

Suggestions for a fix: make Godot case-sensitive with directories. Or at least throw a warning when capitalization does not match.

Edit: these tests were done on macOS.

### PR DESCRIPTION
**Pull request by Kanabenki**
Fixes #23441

### CODE REVIEW

**Comment by aaronfranke** [comment]
In my opinion this should happen on Linux too, so that projects can be cross-compatible. Godot should not allow names that only differ by case, and require names to have the same casing everywhere.

**Comment by Kanabenki** [comment]
It is true that it could be possible to have the same problem on Linux (for example for a project stored on a USB key with a FAT32 filesystem).

**Comment by Kanabenki** [comment]
So, should I remove the macos check so that the warning can be issued on any unix system ?

**Comment by aaronfranke** [comment]
I would appreciate feedback from the Godot devs about this one.

File: drivers/unix/file_access_unix.cpp [hunk 1]
```diff
@@ -104,6 +108,26 @@ Error FileAccessUnix::_open(const String &p_path, int p_mode_flags) {
 		}
 	}
 
+#if defined(OSX_ENABLED) && defined(TOOLS_ENABLED)
+	if (p_mode_flags == READ) {
+		DIR *dir = opendir(path.get_base_dir().utf8().get_data());
+		struct dirent *dir_entry;
+		if (dir) {
+			String base_file = path.get_file();
+			while ((dir_entry = readdir(dir)) != NULL) {
```
Comment (reduz) [inline drivers/unix/file_access_unix.cpp]: This is very inefficient, please look for a different way to do it. Maybe the stat() function is a better fit for this?

**Review by reduz — CHANGES_REQUESTED** [review]

File: drivers/unix/file_access_unix.cpp (hunk 1 above)
Comment (Kanabenki) [reply drivers/unix/file_access_unix.cpp]: The problem is that stat doesn't return the name of the requested file, so there is no way to check the case. I also didn't find a function similar to FindFirstFileW on windows. I'll look into another solution.

**Comment by Kanabenki** [comment]
Thanks for the proposal ! I'm currently stuck trying to find a suitable way of checking for the problem that is compatible with both linux and macos (and haven't had much time to work on it recently), so I would gladly accept help on this one.

**Comment by EmrysMyrddin** [comment]
I've found a way to get the real path on disk of a file. My problem is that it also resolves symlinks and relatives paths (it resolves `.` and `..`)...

I'm a total noob on Godot, but I imagine that including a file can be donne be providing relative links ?

**Comment by EmrysMyrddin** [comment]
Since I dont' have the right to push on your branch and that I haven't used any of your code, I have open a new PR (#24860) to avoid messing with git history.

**Comment by akien-mga** [comment]
Closing as superseded by #24860 which looks like a better option. Still, thanks for the contribution which help get closer to a good solution!

**Pull request closed without merging**